### PR TITLE
Fix fix all behavior

### PR DESCRIPTION
**Pull request by sharwell**
@heejaechang Can you do an initial review of this change? I was able to demonstrate that the change fixed the totally broken behavior of Fix All observed in several cases, but it's unclear whether or not this will negatively impact scenarios I was not able to test.

The root cause appears to be `HasSuccessfullyLoadedAsync` returning false because the `HasSuccessfullyLoadedTransitively` property of the compilation tracker is false. I have no idea how this occurred, but it seems that we do not want to skip the code that finds diagnostics either way.

The bug was most readily reproduced in large solutions (especially e.g. Roslyn.sln), but it's not clear if that is a prerequisite for failure or why it would make a difference.

## Ask Mode

**Customer scenario**

Apply a Fix All operation using the light bulb. The scenario may, but does not always, occur regardless of the scope specified for the Fix All operation.

**Bugs this fixes:**

#17943
#18124

**Workarounds, if any**

The bug does not always occur. Retrying the operation and/or restarting the IDE before retrying the operation in some cases provides a workaround, though the workaround has been observed to be sporadically incomplete.

**Risk**

In the absence of additional work to address the readily-observable performance problems in Fix All operations, this change has a **very high risk** of causing customer machines to either hang or crash. However, the current inconsistent behavior makes the necessary performance evaluations difficult to impossible.

**Performance impact**

This change has been observed to incur substantial overhead when Fix All is used, including very long blocking operations on the UI thread without a wait dialog. However, it appears that the cause of these performance problems was already part of the code and not directly attributed to this change. This change restores consistently correct behavior of Fix All, providing the necessary basis for ongoing performance improvements.

Edit: The performance overhead which caused the UI thread to hang appears to have been caused by code within a `#if DEBUG` section. I'll attempt to narrow this down tomorrow.

**Is this a regression from a previous update?**

This appears to be a regression affecting Visual Studio 2017 to a greater degree than Visual Studio 2015. However, various problems of this form have long been observed in the product.

**Root cause analysis:**

@heejaechang Can you provide more details for the root cause analysis based on my initial comments above?

**How was the bug found?**

Multiple customer and internal reports.

### CODE REVIEW

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs [hunk 1]
```diff
@@ -108,7 +108,7 @@ public IEnumerable<DiagnosticData> ConvertToLocalDiagnostics(Document targetDocu
                         }
 
                         // perf optimization. check whether we want to analyze this project or not.
-                        if (!await FullAnalysisEnabledAsync(project, ignoreFullAnalysisOptions, cancellationToken).ConfigureAwait(false))
```
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs]: no you can't take this one out

**Comment by heejaechang** [comment]
@jinujoseph @mavasani hey, what is you guys opinion on this? basically, project successfully loaded flag indicates whether the project in question we can trust or not (since project failed to loaded correctly can give many incorrect errors due to missing type/wrong overload resolutions and etc). right now, we simply don't return any diagnostic for such project.

in such case, we sometimes shows errors in error list - intellisense build failures, but sometimes we dont (failed skeleton assembly build or missing metadata references or more).

for fix all cases, it might feel like fix all is broken due to missing diagnostics on failed projects.

we can either

1. ignore whether projects are in broken states, and just get diagnostics on those (possibly returning wrong diagnostics at some cases)
2. keep ignore those broken projects and show in fix all popup somewhere that results might not be complete because A, B, C projects are not loaded successfully
3. do (2) but give an option to users to do (1)
4. make the broken projects message a global solution wide thing (like info bar or error list or something) since that actually affects multiple features such as FSA, code fix, fix all, suppression and etc.

there can be more options.

**Comment by heejaechang** [comment]
tagging @dotnet/roslyn-ide as well.

**Comment by heejaechang** [comment]
"Edit: The performance overhead which caused the UI thread to hang appears to have been caused by code within a #if DEBUG section. I'll attempt to narrow this down tomorrow."

this is expected. in debug, we have call to GetDiagnostics for some sanity checks. if you want to test on large solution in Debug bits for fix all, you need to disable that debug check code.

**Comment by CyrusNajmabadi** [comment]
tagging @jasonmalinowski 

@heejaechang The problem here is that when the original work went it, it had the following design:

1. We would mark projects as poisoned if something like an intellisense build failed.
2. we *presumed* that users would get a message saying that they were in such a bad state so that it would be clear to them that there was a root cause causing other issues.

It sounds like '2' is not happening.  We've gotten tons of reports about fix-all failing, but have not heard from anyone (including internals like Kasey) that there's any sort of additional message making it clear waht the problem is.  

Because of this, users are left in a worst-of-all-worlds state.  Fix-all just does nothing, and they have no idea why (or what they can do to even fix it).  

--

Heck, from my perspective, i have no idea what's causing the actual issue.  It may be that our 'poisoning' logic is just busted, and people with totally reasonable projects are getting marked as bad, and then finding at tehy cannot fix-all from that point on.

I am loathe to just band-aid over this problem by saying "ok, we'll just keep going with fix-all".  I think we need to understand why users are getting into this state, and why we're not providing the information for them to understand that tehy're in this bad state.

**Comment by heejaechang** [comment]
@CyrusNajmabadi except intellisense build, I believe we don't report things are busted. 

here is the code that determine whether project is busted or not
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/SolutionState.CompilationTracker.cs,638

we can try to figure out what would be best way to tell users things are busted.

**Comment by heejaechang** [comment]
original design was

1. we set a flag whether we can trust a project or not.
2. we report intellisense build failure to users so that user can fix intellisense build failure. but intellisense build failure is one of cause that can make (1) to happen but that is not only case. for other cases, there was no consensus on whether we should report it or not, how we report it or how to tell users to fix the issue.

**Comment by CyrusNajmabadi** [comment]
Yeah... that's basically a totally unacceptable user experience :)   (and this bug/thread explains precisely why that is).

**Comment by heejaechang** [comment]
yep. Agreed. now we need to figure out how to tell users the issue. and action they can take.

**Comment by CyrusNajmabadi** [comment]
> now we need to figure out how to tell users the issue. 

Or, ideally, figure out what's gone wrong and fix that issue.  It's unclear to me what sort of 'other errors' there are that put us in this state.  Note that we've gotten people reporting fix-all being broken in incredibly simple scenarios.

It's not at all clear to me that we don't just have some sort of low level bug that we're just masking with this (or there's just a bug in the poisoning logic) and that low-level bug is just going to stay around causing problems.

In other words, are we just proposing treating the symptoms?  Or are we addressing the actual root problem?

**Comment by mavasani** [comment]
I think `HasSuccessfullyLoadedAsync` check is not something that the diagnostic service should do - some clients (such as FixAll) may really want to get diagnostics even in broken state. I think any such filtering should happen at a client level (in this case error list) OR the diagnostic service APIs should take in a flag whether or not to provide broken project diagnostics or not.

**Comment by heejaechang** [comment]
project is in broken state if any of metadata reference is missing or it has skeleton assembly which is failed to build except intellisense build failed.

**Comment by heejaechang** [comment]
@mavasani sure we can do that way as well. we can either add flag like suppression or just assume features know about it and assume it does right thing.

...

by the way, I am not sure whether Fix all completely ignore it is right thing since it can return wrong diagnostics. my preference will be at least fix all pop up telling users such such projects are in broken state.

**Comment by CyrusNajmabadi** [comment]
> I think any such filtering should happen at a client level (in this case error list) OR the diagnostic service APIs should take in a flag whether or not to provide broken project diagnostics or not.

I'm still concerned that this will still jsut be masking some underlying issue.  I've seen people repro this with literally just console apps.  I think we've identified that we've gone off the rails, and we don't want the experience (like around diagnostics in the error list) to be terrible.  But we haven't identified why we've gone off the rails.  

We believe there are cases we can't handle at all (like intellisense build failure), and that that is ok to move us into this state.  HOwever, we have no code that validates that the problems we're seeing in fix-all stem from that problem or if they stem from some other root cause.  Without knowing that we don't know if there's some serious underlying issue that we're just not adequately addressing.

**Comment by heejaechang** [comment]
missing metadata and failed skeleton assembly is also something we can't control.

**Comment by CyrusNajmabadi** [comment]
> or just assume features know about it and assume it does right thing.

> by the way, I am not sure whether Fix all completely ignore it is right thing since it can return wrong diagnostics

I agree.  I'm would not want fix-all doing *anything* if hte project is in a busted state.  It just means we can corrupt user code since we'll be making bad decisions on incomplete/inaccurate information.

**Comment by CyrusNajmabadi** [comment]
> missing metadata and failed skeleton assembly is also something we can't control.

We've seen repros containing no skeleton assemblies.  

Could you clarify what the 'missing metadata' issue is?  I would assume if there was missing metadata then the user would get an error saying something was wrong and that metadata could not be loaded.  But we've seen *in person* fix-all problems where the error list was clear.  

Thanks!

**Comment by heejaechang** [comment]
I shared the code
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/SolutionState.CompilationTracker.cs,638

this is the code that determine whether a project is in good state or not. code is probably most precise way to tell what the code think missing metadata is.

**Comment by mavasani** [comment]
>  I'm would not want fix-all doing anything if hte project is in a busted state

In that case it would be weird for the end user to see their code fixes working fine, but FixAll being a no-op. We should either hide code fixes completely (which I guess is unacceptable), hide the FixAll hyperlinks on light bulb (which is again confusing), or give the popup that HeeJae mentioned so that they can review the FixAll preview changes dialog carefully before committing or even just show a **noticeable** warning message at the top of FixAll preview changes dialog about this issue. I don't think we can keep blocking FixAll completely for such cases, without giving any justification or control to the user.

**Comment by CyrusNajmabadi** [comment]
> In that case it would be weird for the end user to see their code fixes working fine, but FixAll being a no-op. 

I don't actually understand how this system works.  If you guys aren't producing diagnostics when the project is in a busted state, how are normal fixes even working?

**Comment by heejaechang** [comment]
we decide to ignore the flag for open/active files since context is at user's face so they most likely can see whether it is right or wrong. and decide whether they will run code fix or not.

fix all's problem is it includes ones who contexts are not obvious..

by the way, when one says fix all not working, I assumed it worked partially. I mean the flag probably won't affect the fix all features as whole. but some projects that failed, diagnostics won't be reported, so those proejcts won't have fix applied. but other projects that are successfully loaded, will have fix applied.

**Comment by CyrusNajmabadi** [comment]
> by the way, when one says fix all not working, I assumed it worked partially. I mean the flag probably won't affect the fix all features as whole

In the case of fix-all, all the bugs we've sen reported are just that *nothing* is updated when 'fix all' is invoked.  This include when the user just picks "fix all in document".  Even in that case, no fixing occurs.  We got another case of this just a couple of days ago, and we've been hearing the reports on this for nearly a year now.

**Comment by heejaechang** [comment]
that issue's cause can't be same as this fix though. that must be other issue. if it is related to OpenFileOnly IDE analyzer, that is due to known issue where fix all doesn't work for OpenFileOnly analyzers. 

I know the cause, just didn't have time to fix it yet.

**Comment by sharwell** [comment]
I still need to go through all the feedback, but I did want to point one item out:

>  It may be that our 'poisoning' logic is just busted, and people with totally reasonable projects are getting marked as bad, and then finding at tehy cannot fix-all from that point on.

I have a suspicion this is occurring. In particular, I observed that a "partial" compilation state was getting created, and that state always has the `HasSuccessfullyLoadedTransitively` forced to false.

Also, keep in mind that fix all needs to work at minimum as "best effort" in broken project scenarios. It is acceptable to indicate to the user that the results may be inaccurate, but it is not acceptable to disable the feature altogether. It is as vital to the editing experience as code completion and formatting, and the ramifications of disabling those in broken projects is well-established.

**Comment by sharwell** [comment]
@heejaechang 

> Agreed. now we need to figure out how to tell users the issue. and action they can take.

Analyzers are expected to work in broken code scenarios, both locally (e.g. invalid syntax in a file) and globally (e.g. unresolved references). Code fixes are expected to either disable *themselves* (logic in the analyzer or fix) or behave sanely in these scenarios. Failure to achieve these is a report that we on StyleCop Analyzers treat as a bug, and we have many regression tests targeting such scenarios specifically. Given this, users would not need a dedicated indicator of project state in order to have a great overall experience.

**Comment by sharwell** [comment]
> by the way, when one says fix all not working, I assumed it worked partially. I mean the flag probably won't affect the fix all features as whole. but some projects that failed, diagnostics won't be reported, so those proejcts won't have fix applied. but other projects that are successfully loaded, will have fix applied.

I reproduced the behavior of #17943 several times during testing for the Remove Unused Usings analyzer attempting to Fix All in Document. A failure at this scale is essentially a total failure of Fix All (no fixes suggested for any Fix All action attempted for any selected scope).

**Comment by sharwell** [comment]
> ignore whether projects are in broken states, and just get diagnostics on those (possibly returning wrong diagnostics at some cases)

This is my preferred approach. This is also the approach StyleCop Analyzers is expecting the diagnostic engine to be using.

**Comment by CyrusNajmabadi** [comment]
@sharwell Do you have a repro that's even mildly reproducible?  If so, that is something that would be great for root causing what's happening here.  For example, we could put in logging to see precisely why we were not able to fully load the project properly.

**Comment by sharwell** [comment]
@CyrusNajmabadi 

> Do you have a repro that's even mildly reproducible?

This was sufficiently hard to reproduce that I actually preserved my original A/B testing commit where I used a debugger to verify the fix in a live instance where the bug occurred. See the first commit of this PR.

Sometimes the issue is reproduced for Fix All in Document for the Remove Unused Usings analyzer. When this occurs, attach a debugger to the instance as it tends to stay in this state once it's entered. In Roslyn.sln, it occurred regularly enough that it didn't take too much time to find.

**Comment by CyrusNajmabadi** [comment]
If we can use Roslyn.sln to repro this, it sounds like we could add appropriate logging to help us root cause things.  That would help enormously in root causing the underlying issue which the diagnostics engine is trying to then protect itself from.

**Comment by sharwell** [comment]
@CyrusNajmabadi Even if we figure out why it's incorrectly treating the project as incomplete and disabling the fix all, it doesn't change the fact that we want fix all scenarios to work *even if the project really is incompletely loaded*.

**Comment by CyrusNajmabadi** [comment]
>  it doesn't change the fact that we want fix all scenarios to work 

I didn't say otherwise :)

My concern is around us masking an underlying issue and that issue persisting and affecting other parts of the system.   In this regard, i want us to make sure we're trying to fix the root problem, and not *just* fixing the most notable symptom of it.

**Comment by CyrusNajmabadi** [comment]
> I have a suspicion this is occurring. In particular, I observed that a "partial" compilation state was getting created, and that state always has the HasSuccessfullyLoadedTransitively forced to false.

If this is the case it is *very* bad, and we need to track that down as it is breaking invariants of our system.  You should only be getting a partial compilation state if you ask for "GetDocumentWithFrozenPartialSemanticsAsync", and that should only ever really be used in the completion scenario (and maybe a couple of other cases).  If tehse states are leaking out then something really bad is happening.

**Comment by heejaechang** [comment]
"I observed that a "partial" compilation state was getting created, and that state always has the HasSuccessfullyLoadedTransitively forced to false."

oh. this might be the root cause. codefix or refactoring should never use partial solution.

**Comment by heejaechang** [comment]
@sharwell I agree that code fix/refactoring should work in a situation where solution contains errors. but that is different than error code fix is invoked upon itself is wrong.

the flag indicates we can't trust errors that belong to the project the flag is set. which is different than errors I am invoking code fix on is probably okay but the project itself doesnt compile due to some errors.

**Comment by Pilchie** [comment]
>if it is related to OpenFileOnly IDE analyzer, that is due to known issue where fix all doesn't work for >OpenFileOnly analyzers.
>
>I know the cause, just didn't have time to fix it yet.

@heejaechang - can you elaborate on this?  It sounds like there are two issues that are getting conflated here.

1. The issue you mention, which I have no insight into.
2. The issue this PR attempts to deal with - that Fix All doesn't work if any metadata references are missing.  As stated above, I think the current experience there is very bad.  I'd like to propose that we merge this for 15.3, while we figure out the right way to expose the problems.  You say that the errors could be wrong in this case.  My guess is that they are wrong in one of a few ways though:

a. Extra errors about types and members not being found because of the missing references.
b. Expected errors not being reported, because semantic analysis can't progress far enough to find them.
c. I know usings are marked as unused if they don't resolve to an actual namespace.  

In the case of a, I don't think we have (or expect) any fix-alls for this category of error, so I'm not that worried about it.

In the case of b, I think allowing fix all to work is fine, since it will just result in not all instances being fixed.

For c, maybe we should make the fix not remove unused usings that have that specific error on them.

Are there other concerns?

It is probably true that we want to move this change to something specific to Fix All though, as not reporting all of those errors is one of the ways we cut down on cascading errors.

Thoughts?

**Comment by heejaechang** [comment]
@Pilchie if we check this in, we will again have error list having thousands of errors when solution/project is  in broken state.

**Comment by mavasani** [comment]
I think we should not block FixAll on this flag. Until we fix the root cause of the why this flag is getting set for unexpected cases, we should special case FixAll code path to ignore this flag.

**Comment by heejaechang** [comment]
sure, we could special case it for fix all, but this change won't be it. since this same code will be shared by both fix all and all other cases.

when special case it for fix all, one should consider caching, error list, lightbulb, squiggles, fix all and close file cases. right now, every one share 1 back end which make things simpler.

once we special case it for fix all, one need to make sure all works as expected and only special case fix all cases.

**Comment by Pilchie** [comment]
@heejaechang That's why I said:

>It is probably true that we want to move this change to something specific to Fix All though, as not reporting all of those errors is one of the ways we cut down on cascading errors.

😄 

Can you elaborate on the Open Files Only issue you mentioned above though?

**Comment by heejaechang** [comment]
@Pilchie the call path simply (or in certain cases, can't remember exactly on top of my head) doesn't include any analyzer that is marked as open file only. so fix works but no fix all work for open file only analyzers.

**Comment by Pilchie** [comment]
Is there a bug number, or some repro steps that we could use?  I'd like to get that fixed ASAP.

**Comment by heejaechang** [comment]
@mavasani do you remember that bug? was it vso bug?

**Comment by heejaechang** [comment]
I am doing some work around this. so, let me spend some time to fix this issue.

**Comment by heejaechang** [comment]
@mavasani @Pilchie https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?id=399140&_a=edit&triage=true

thank you manish for finding the bug for me. this is basically same issue with same root cause

**Comment by sharwell** [comment]
Superseded by #19714 and #19637.
